### ISSUE TITLE
False-negative: Member names with underscores

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17842]
From: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case
> In very rare circumstances (for example, multipart version numbers), you may need to use underscores to separate adjacent numbers, since numbers do not have upper and lower case variants.

Google style is not exact. On one hand it says `_` can be used in test methods to combine sentences and on another hand in general CamelCase explanation, `_` is allowed for numbering suffix in all names(except package and generic names).

Valid uses of underscore :
```
1. guava33_4_5
2. guavaVersion33_4_5 
3. guavaversion33_4_5 (same as 1.)
```

Invalid uses of underscore:
```
1. guava_33_4_5_ (underscore at the end)
2. guava_33_4_5   (underscore between digit and letter)
3. guava_version33_4_5 (underscore between lowercase character sequences)
```

Some false-negatives in member names left to address from #17708  

```java
$ cat NamesWithUnderscores.java
/** some javadoc. */
public class NamesWithUnderscores {
  int gradle9_5_1 = 0;

  int gradle9_5_1_ = 0; // ok

  int gradle_9_5_1 = 0; // Expected violation, _ between digit and letter

  int jdk_9_0_392 = 0; // Expected violation, _ between digit and letter

  int testing_01231 = 0; // Expected violation, _ between digit and letter
}
```

```
$ java -jar checkstyle-11.1.0-all.jar -c google_checks.xml NamesWithUnderscores.java 
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/NamesWithUnderscores.java:5:7: Member name 'gradle9_5_1_' must match pattern '^(?![a-z]$)(?![a-z][A-Z])[a-z][a-zA-Z0-9]*(?:_[0-9]+)*$'. [MemberName]
Audit done.
```

we probably need a better approach than Regex as current regex has already grown very complex.

https://github.com/checkstyle/checkstyle/blob/5d52248c85b0fc37ddd786ddb5354e332f8a54b8/src/main/resources/google_checks.xml#L254-L256

if special check is needed, it should first determine the type of name - regular or name with numbering and then it should validate the name against the appropriate regex format, which will differ for each type of naming

**Comment by romani** [comment]
current Check is https://checkstyle.sourceforge.io/checks/naming/membername.html 
We should create `GoogleMemberName` Check to validate numbers in names as Google want, no properties should be in such Check, as it will handle only Google requirements, not a general use.

**Comment by vivek-0509** [comment]
i am on it

**Comment by vivek-0509** [comment]
Hi @romani ,

Before proceeding further with the implementation of `GoogleMemberNameCheck`, I would like to clarify the valid use cases of underscores in non-constant field names.

While going through the PR and related discussions, I found some ambiguity around when underscores are considered acceptable. Below is my current understanding, and I would appreciate confirmation or correction.

## My current understanding

### Clearly valid
- `guava33_4_5`  
  *Digit immediately before underscore, used only to separate numeric version parts.*
- `kotlin1_9_24`  
  *Word ending with digit, followed by numeric version suffix.*
- `kotlinVersion1_9_24`  
  *CamelCase name ending with digit, followed by numeric version suffix.*

### Clearly invalid
- `gradle_9_5_1`  
  *Letter immediately before underscore.*
- `kotlin_version1_9_24`  
  *Underscore used to separate words.*
- `mField`, `pValue`  
  *Hungarian notation style prefixes.*

### ? Unclear / need clarification
- `guavaSomething_version33_4_6`  
  *CamelCase followed by underscore and a word, then numeric version.*
- `guavaSomething_versionSomething33_4_6`  
  *Should underscore between CamelCase segments be valid or invalid?*

## Questions needing clarification

1. Is underscore usage **only valid between digits** (for example `33_4_5`), and invalid in all other positions?

2. **In general, what are the explicitly valid and explicitly invalid underscore patterns that this check should enforce?**

**Comment by romani** [comment]
>guavaSomething_version33_4_6
guavaSomething_versionSomething33_4_6

Violation for both.
Expected naming to comply Google rules:
guavaSomethingVersion33_4_6
guavaSomethingVersionSomething33_4_6

>Is underscore usage only valid between digits (for example 33_4_5), and invalid in all other positions?

Underscore is used as substitution for ".". 33.4.5 to 33_4_5 .
Only this substitution is allowed, not others.

**Comment by vivek-0509** [comment]
@romani 
Thanks for the clarification .

**Comment by vivek-0509** [comment]
Hey @romani,

Per [Google Style Guide 5.2.5](https://google.github.io/styleguide/javaguide.html#s5.2.5-non-constant-field-names), non-constant field names "(static or otherwise)" should follow lowerCamelCase. Currently google_checks.xml doesn't include StaticVariableNameCheck, so static non-final fields are not validated. Since GoogleMemberNameCheck is meant to enforce Google's naming rules for fields, should it also validate static non-final fields, or only instance fields?

**Comment by vivek-0509** [comment]
@romani ping

**Comment by romani** [comment]
Google special module GoogleNonConstantFieldName is activated in Google style config
